text/dependencies/rfcs/0000-dependency-management-phase-two.md

```diff
@@ -0,0 +1,195 @@
+`@paketo-buildpacks/dependencies-maintainers` to the owners of the
+`actions/dependency` and `implementation/.github/workflows/<all
+dependency-related workflows>`.
+
+
+Each of the step s outlined in the overview will be converted into workflow
+steps.
+
+Workflow 1:
+1. Picking up new versions with buildpack code will be a workflow step, which
+   runs on a timer at least once per day.
+
+2. Figuring out which versions will be added to the `buildpack.toml` will be
+   rolled into an action that outputs versions. It will leverage parts of the
+   existing `jam update-dependencies` logic.
+
+3. For individual versions, a Github Actions [`matrix
+   strategy`](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#using-a-matrix-strategy)
+   will be used to iterate over each version to be added, to generate metadata
+   using the code from within the buildpacks.
+
+   The metadata generation code from the buildpack will contain logic to spit
+   out "variants" for each dependency version, depending on the stacks
+   (OS/architectures) supported by the buildpack.
+
+   For example, if the dependency is used from source but has two different
+   artifacts for ARM64 and AMD64 variants, and the buildpack supports both
+   architectures, then the metadata generation code should spit out two batches of
+   metadata for each new version. One with the AMD64 upstream source URI/SHA256, and
+   the other with the ARM64 upstream URI/source SHA256.
+
+5. In the same matrix loop, dependency compilation will be kicked off with the
+   code from the buildpacks on the basis of whether the metadata contains the
+   `SHA256` and `URI` of the dependency.
+
+6. A smoke test will be run against the dependency in both the
+   compiled/non-compiled cases via a step that simply runs the test from
+   `<buildpack>/dependency/test`.
+
+7. (If compiled) The dependency will be uploaded via an action
+
+8. (If compiled) The metadata is updated with the SHA256 and GCP bucket URI in
+   a workflow step
+
+9. Updating the `buildpack.toml` with the new versions and metadata will also
+   be delegated to a new action
+
+10. A pull request will be opened with the changes to the `buildpack.toml`. The
+    usual suite of integration and unit tests will be run against the pull
+    request as in all other cases.
+
+
+### Updating Dependencies Manually
+All of these steps will run as a workflow with the option for a manual
+dispatch, but can be run easily manually on a local system.  Additionally, we
+could eventually create a script inside of the `<buildpack>/scripts` directory
+to perform the main steps of the `buildpack.toml` update process locally. This
+will replace the need for the `jam update-dependencies` command that the
+automation uses currently. Since dependency-related logic will live alongside
+the buildpack, it makes more sense for manual dependency-update scripts to live
+there as well in order to leverage code.
```

Comment:
I am a little curious why we would want to write a new script instead of just modifying the functionality of `jam update-dependencies` to do whatever that script would do, or even potentially adding some new command if you want to keep the previous generations functionality.

Reply:
I guess I just think it would be better to keep as many buildpack development tool in `jam` as possible to hopefully help spread this way of doing dependencies but I am not tied to the idea.

Reply:
A lot of the functionality of `jam update-dependencies` is spread out into smaller, composable pieces so that they can be plugged in with other functionality in order to support a bunch of different use cases.

I'd say that the "Assemble" step (see step 9 in [the overview](https://github.com/paketo-buildpacks/rfcs/blob/dependency-management-step-two/text/dependencies/rfcs/0000-dependency-management-phase-two.md#new-plan-overview)) might be the best candidate for pushing into `jam update-dependencies` instead of a Github action. It would be a command that just knows how to update the `buildpack.toml` given metadata. I'm not particularly married to it being an action or a `jam` command really

Reply:
I’d say, better to be an action if that step should run in a container, better to put in `jam` if it’ll “just work” on any CLI users local workstation, given the appropriate inputs.

Reply:
alright seeing as two tooling maintainers are both on board with the metadata update step as a part of `jam`, I have updated the RFC to reflect that. It does not need to run inside of a container specifically, so it probably doesn't need to be an action.